daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Util.scala

```diff
@@ -84,4 +84,69 @@ object Util {
   val CPTrue = CPPrimCon(PCTrue)
   val CPFalse = CPPrimCon(PCFalse)
 
+  type Graph[X] = Map[X, Set[X]]
+
+  // Topologically order the vertices of an abstract Graph.
+  // If the `graph` is a directed acyclic graph returns a list of its vertices in topological order as `Right`
+  // otherwise returns a list of vertices that form a cycle as `Left`
+  def topoSort[X](graph: Graph[X]): Either[List[X], List[X]] = {
+
+    var white = graph.keySet
+    var black = graph.values.foldLeft(InsertOrdSet.empty[X])(_ | _.filterNot(white))
+    def gray(x: X): Boolean = !white(x) && !black(x)
+
+    def visitSet(xs: Set[X]): Option[X] = xs.foldLeft(Option.empty[X])(_ orElse visit(_))
+
+    def visit(x: X): Option[X] =
+      if (black(x))
+        None
+      else if (!white(x))
+        Some(x)
+      else { white -= x; visitSet(graph(x)) } orElse { black += x; None }
+
+    def buildCycle(curr: X, start: X, list: List[X] = List.empty): List[X] = {
+      val next = graph(curr).find(gray).getOrElse(throw new UnknownError)
+      if (next == start)
+        curr :: list
+      else
+        buildCycle(next, start, curr :: list)
+    }
+
+    visitSet(graph.keySet).fold[Either[List[X], List[X]]](Right(black.toList))(x =>
+      Left(buildCycle(x, x)))
+  }
+
+  // Return the `pkgIds` and all its dependencies in topological order.
+  // A package undefined w.r.t. the function `packages` is treated as a sink.
+  def dependenciesInTopologicalOrder(
+      pkgIds: List[Ref.PackageId],
+      packages: Ref.PackageId PartialFunction Package,
+  ): List[Ref.PackageId] = {
+
+    @tailrec
+    def buildGraph(
+        toProcess0: List[Ref.PackageId] = pkgIds,
+        seen0: Set[Ref.PackageId] = pkgIds.toSet,
+        graph0: Graph[Ref.PackageId] = HashMap.empty,
```

Comment:
I'd prefer passing these in in the call below rather than setting default arguments.